public/app/features/alerting/state/AlertingQueryRunner.ts

```diff
@@ -0,0 +1,208 @@
+  LoadingState,
+  PanelData,
+  rangeUtil,
+  TimeRange,
+} from '@grafana/data';
+import { BackendSrvRequest, FetchResponse, toDataQueryError } from '@grafana/runtime';
+import { BackendSrv, getBackendSrv } from 'app/core/services/backend_srv';
+import { preProcessPanelData } from 'app/features/query/state/runRequest';
+import { GrafanaQuery } from 'app/types/unified-alerting-dto';
+import { getTimeRangeForExpression } from '../unified/utils/timeRange';
+import { isExpressionQuery } from 'app/features/expressions/guards';
+
+export interface AlertingQueryResult {
+  frames: DataFrameJSON[];
+}
+
+export interface AlertingQueryResponse {
+  results: Record<string, AlertingQueryResult>;
+}
+export class AlertingQueryRunner {
+  private subject: ReplaySubject<Record<string, PanelData>>;
+  private subscription?: Unsubscribable;
+  private lastResult: Record<string, PanelData>;
+
+  constructor(private backendSrv = getBackendSrv()) {
+    this.subject = new ReplaySubject(1);
+    this.lastResult = {};
+  }
+
+  get(): Observable<Record<string, PanelData>> {
+    return this.subject.asObservable();
+  }
+
+  run(queries: GrafanaQuery[]) {
+    if (queries.length === 0) {
+      const empty = initialState(queries, LoadingState.Done);
+      return this.subject.next(empty);
+    }
+
+    this.subscription = runRequest(this.backendSrv, queries).subscribe({
+      next: (dataPerQuery) => {
+        const nextResult: Record<string, PanelData> = {};
+
+        for (const [refId, data] of Object.entries(dataPerQuery)) {
+          const previous = this.lastResult[refId];
+          nextResult[refId] = setStructureRevision(data, previous);
+        }
+
+        this.lastResult = nextResult;
+        this.subject.next(this.lastResult);
+      },
+      error: (error) => console.error('PanelQueryRunner Error', error),
+    });
+  }
+
+  cancel() {
+    if (!this.subscription) {
+      return;
+    }
+    this.subscription.unsubscribe();
+
+    const nextResult: Record<string, PanelData> = {};
+    let loading = false;
+
+    for (const [refId, data] of Object.entries(this.lastResult)) {
+      if (data.state === LoadingState.Loading) {
+        loading = true;
+      }
+
+      nextResult[refId] = {
+        ...data,
+        state: LoadingState.Done,
+      };
+    }
+
+    if (loading) {
+      this.subject.next(nextResult);
+    }
+  }
+
+  destroy() {
+    if (this.subject) {
+      this.subject.complete();
+    }
+
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+  }
+}
+
+const runRequest = (backendSrv: BackendSrv, queries: GrafanaQuery[]): Observable<Record<string, PanelData>> => {
+  const initial = initialState(queries, LoadingState.Loading);
+  const request = {
+    data: { data: queries },
+    url: '/api/v1/eval',
+    method: 'POST',
+    requestId: uuidv4(),
+  };
+
+  const runningRequest = getBackendSrv()
+    .fetch<AlertingQueryResponse>(request)
+    .pipe(
+      map(mapToPanelData(initial)),
```

Comment:
Maybe add a takeUntil  here that handles the cancellation?

Reply:
We have a `takeUntil` on line 119.

Reply:
🙈  I missed that

Reply:
Aha, nope, that is not what I meant, that takeUntil you're mentioning is just so we receive an initial result within 200ms. I meant that the cancel function should cancel this call to fetch.

Reply:
aha, got it! I will add that. Super good feedback.

Reply:
hmm, isn't this already covered by the `cancelNetworkRequestsOnunsubscrive` ?

Reply:
Good point, but that is mainly used for data sources that haven't migrated to RxJs and rely on strings. I think it's a lot safer to cancel your own observables when you can.